### PR TITLE
Test pvsystem.singlediode against precisely generated IV curves

### LINKED ISSUE #411: Test singlediode() implementations using a "gold" function and data set
#409 Suggests Bishop [1988] as a truly explicit method to calculate i-from-v. Bishop's method is not subject to Newton convergence failure or to LambertW argument "blow up", handles ideal device parameters without extra logic, has no dependency on scipy, and naturally vectorizes via standard numpy functions' vectoization. The trade-off is that the explicit device terminal current computation occurs at the diode junction voltage (V_d := V + I R_s), not the device terminal voltage. (The device terminal voltage is "back-calculated" as V = V_d - I Rs.) Thus, additional computation work must be done to determine currents at specified terminal voltages, inc. I_sc @ V_sc=0, I_mp @ V_mp, and I_oc=0 @ V_oc.

The Bishop method is a very good candidate for a "gold" function implementation for testing the accuracy and speed of [various implementations of `i_from_v()`](#410) that pvlib may wish to offer users or refactor to. By inverting the dataset, it could also be used to test `v_from_i()` implementations.
```python
singlediode(photocurrent, saturation_current, resistance_series, resistance_shunt, nNsVth, ivcurve_pnts=None)
```

I think that after this `bishop88` function is implemented (presumably by #409), we should create and commit a corresponding "gold" dataset that we can use for better testing, refactoring, and optimization of the `singlediode()` function. I think we should try to cover a large number (all?) of the devices in the CEC database over a combination of temperature and irradiance ranges (perhaps the IEC 61853-1 matrix?). This dataset should also be valuable to the greater PV modeling community who wish to test/contribute new computational algorithms.

Outstanding questions:
(1) What should the voltage range be? Specifically, the diode voltages will give some terminal voltages less than zero, and how much beyond V_oc should we try to extend?
(2) Can/should we also include I_sc, I_x, I_mp, I_xx, and V_oc points, and what about the slope at I_sc and slope at V_oc?
(3) Does python have an interval analysis package up to the task of verifying our computations, e.g., [intPy](https://pypi.python.org/pypi/IntPy)?
(4) Is it feasible to run such a comprehensive test suite on every PR?

cc @mikofski @adriesse

### PR DESCRIPTION
**Pull request by reepoi**
- [x] Closes #411
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - ~~[ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/reference) for API changes.~~
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

Generated IV curves are added to `pvlib/data` to test the precision of `pvsystem.singlediode`, `pvsystem.v_from_i`, and `pvsystem.i_from_v`. The parameters of the single diode equation that generated each IV curve are in CSV files, and the IV curve data are in JSON files following [this JSON schema](https://github.com/cwhanse/ivcurves/blob/main/ivcurves/ivcurve_jsonschema.json). `i_x` and `i_xx` are added in addition to the fields of the JSON schema since they are also calculated by `pvsystem.singlediode`.
The generated IV curve data are calculated using the functions in the [ivcurves repository](https://github.com/cwhanse/ivcurves). These functions use the arbitrary precision math library [mpmath](mpmath.org). The data may be reproduced by checking out the code of [this commit of ivcurves](https://github.com/reepoi/ivcurves/tree/19c0269884fbb5f35c385b61923175bfbc961a13) and running the command
```
python3 ivcurves/precise.py --save-json test_sets
```
The data are in the files `test_sets/case1.json` and `test_sets/case2.json`.

### CODE REVIEW

File: pvlib/pvsystem.py
Comment (cwhanse) [inline pvlib/pvsystem.py:2879]: These changes should be reverted. `singlediode.bishop88` returns the current and voltage vectors. The returned voltage vector is not equal to the input `vd` which is the diode voltage.

Arguably, the returned voltage could be constructed in the way proposed here (compute voltage, not diode voltage, and find the corresponding currents). But that should be discussed in a separate issue.

Comment (reepoi) [reply pvlib/pvsystem.py]: I reverted these changes and created issue #1578 for this. The test cases for the `brentq` and `newton` methods check the accuracy of the currents returned at the voltages of `vd`, so they will not pass. Since these unit tests depend on what voltage values are returned by `pvsystem.singlediode`, I think they should be removed for now and added when issue #1578 is resolved.

**Comment by reepoi** [comment]
I updated the description of this pull request. The test cases added now test:
- `pvsystem.singlediode` with `ivcurve_pnts=None` and all options for `method`: checks all the returned values.
- `pvsystem.singlediode` with `ivcurve_pnts=100` and `method='lambertw': checks only the returned current and voltage values.
- `pvsystem.v_from_i` and `pvsystem.i_from_v` with all options for `method`: checks the returned voltage or current values.

The tests for `pvsystem.v_from_i` and `pvsystem.i_from_v` indirectly test the precision of `pvsystem.singlediode` with methods `brentq` and `newton` when `ivcurves_pnts=100` since it they all eventually call `pvsystem.bishop88`.

**Comment by cwhanse** [comment]
Ready for review. @markcampanelli would appreciate your input if this closes #411

File: pvlib/data/precise_iv_curves_parameter_sets1.csv
Comment (markcampanelli) [inline pvlib/data/precise_iv_curves_parameter_sets1.csv:2]: Because this is a Python repo, you might consider indexing from 0, not 1.

File: pvlib/data/precise_iv_curves1.json
Comment (markcampanelli) [inline pvlib/data/precise_iv_curves1.json:11]: Because this is a Python repo, you might consider indexing from 0, not 1.

File: pvlib/tests/test_singlediode.py
Comment (markcampanelli) [inline pvlib/tests/test_singlediode.py:115]: For better traceability, I would think that the choice of values for these constants should documented as part of the files containing the parameters used to precisely compute the curves.

For comparison, `scipy==1.8.1` gives:

scipy.constants.value("Boltzmann constant") = 1.380649e-23 J/K
scipy.constants.value("elementary charge") = 1.602176634e-19 C

File: pvlib/tests/test_singlediode.py
Comment (markcampanelli) [inline pvlib/tests/test_singlediode.py:158]: I'm super curious how these tolerances might be tightened for 'lambertw', 'brentq', vs. 'newton'.

**Review by markcampanelli — CHANGES_REQUESTED** [review]
Thanks for this submission. Generally looks good.

The big thing I would like to see is some better documentation in this repo (mostly for posterity) about how the precise I-V data points were generated elsewhere. For example, how did you choose the precision cutoff using `mpmath` with respect to the precision used in the `pvsystem.singlediode` calculations? The choice of constants in the upstream calculation could also have a better "paper trail".

File: pvlib/tests/test_singlediode.py
Comment (markcampanelli) [reply pvlib/tests/test_singlediode.py:115]: Also see https://github.com/pvlib/pvlib-python/issues/483

**Comment by markcampanelli** [comment]
> Ready for review. @markcampanelli would appreciate your input if this closes #411

I think this addresses https://github.com/pvlib/pvlib-python/issues/411, so that can be closed once this merges.

I have closed https://github.com/pvlib/pvlib-python/pull/426, which this substitutes.

File: pvlib/tests/test_singlediode.py
Comment (reepoi) [reply pvlib/tests/test_singlediode.py]: Ok, I updated these constants to match The SI Brochure (https://www.bipm.org/en/publications/si-brochure) and cited that document in the docstring. In that docstring, I also added the precision cutoff of mpmath and more detail about the precision of the IV curve data. I avoided mentioning [cwhanse/ivcurves](https://github.com/cwhanse/ivcurves), but maybe we should reference it. There is much more documentation and code that would help someone understand how the IV curve data was generated.

Also, if we use `scipy.constants`, pvlib may need to update its minimum SciPy version requirement because Boltzmann's constant changes depending on the SciPy version. The value for SciPy 1.9.3 is `1.602176634e-19 C`, which matches the value hard-coded here; however, the value is `1.6021766208e-19 C` for SciPy 1.2.0. It changes in SciPy 1.4.0, and pvlib's minimum version of SciPy is 1.2.0.

Comment (cwhanse) [reply pvlib/tests/test_singlediode.py]: @pvlib/pvlib-maintainer I think we should advance scipy version and use the 2019 values, which are defined to be _exact_ rather than with uncertainty.

File: pvlib/tests/test_singlediode.py
Comment (cwhanse) [reply pvlib/tests/test_singlediode.py]: And I'm curious why there are test failures only on ubuntu and python <=3.9 for the 'lambertw' case. I suspect something buried fairly deep behind `scipy.special.lambertw` which we won't solve.

Comment (reepoi) [reply pvlib/tests/test_singlediode.py]: For [this failed test](https://github.com/pvlib/pvlib-python/actions/runs/3879773557/jobs/6617268775), the 30th `i_xx` value differs by `1e-6`. All of the tests related to this pull request pass with `atol=1e-6` for the `i_xx` value.

Comment (cwhanse) [reply pvlib/tests/test_singlediode.py]: @reepoi I think there's a mistake in the precise curve library somewhere.

**Comment by cwhanse** [comment]
> The current test failures are only on linux, only occur for numpy>=1.22.0, and have to do with precision checks, so I suspect this is due to the increased SIMD functionality [added in numpy 1.22.0](https://numpy.org/doc/stable/release/1.22.0-notes.html#vectorize-umath-module-using-avx-512).
> 
> In the past we have relaxed test tolerances to accommodate the slight numerical variation, but in this specific case that seems rather self-defeating :) 

My perspective is that the test tolerance is a measure of the pvlib functions' accuracy. Six digits (worst case) is at least useful for energy simulations and may also help provide context for other use cases. At least it establishes a benchmark from which we can measure improvements.

**Comment by kanderso-nrel** [comment]
@cwhanse a lot has happened in the https://github.com/cwhanse/ivcurves repository in the last couple months.  Are any corresponding updates required in this PR?

**Comment by cwhanse** [comment]
We may need to replace the json files

**Comment by reepoi** [comment]
I regenerated the JSON files using the latest version of the https://github.com/cwhanse/ivcurves repository and there is no difference between them and the JSON files in this PR.

**Review by kanderso-nrel — APPROVED** [review]
A few minor docs suggestions but otherwise LGTM.  Thanks @reepoi!

**Review by cwhanse — APPROVED** [review]

**Comment by kanderso-nrel** [comment]
Thanks @reepoi for the contribution and @markcampanelli and @cwhanse for the reviews!

**Pull request merged**
